src/commons/editor/Editor.tsx

```diff
@@ -50,529 +64,249 @@ type StateProps = {
+ */
+const selectMode = (chapter: number, variant: Variant, library: string) => {
+  HighlightRulesSelector(chapter, variant, library, Documentation.externalLibraries[library]);
+  ModeSelector(chapter, variant, library);
+};
 
-    this.handleVariableHighlighting();
+const makeHandleGutterClick = (
+  handleEditorUpdateBreakpoints: DispatchProps['handleEditorUpdateBreakpoints']
+) => (e: AceMouseEvent) => {
+  const target = e.domEvent.target! as HTMLDivElement;
+  if (
+    target.className.indexOf('ace_gutter-cell') === -1 ||
+    !e.editor.isFocused() ||
+    e.clientX > 35 + target.getBoundingClientRect().left
+  ) {
+    return;
   }
 
-  public componentWillUnmount() {
-    if (this.ShareAce !== null) {
-      // Umounting, closing websocket
-      this.ShareAce.WS.close();
-    }
-    this.ShareAce = null;
+  // Breakpoint related.
+  const row = e.getDocumentPosition().row;
+  const content = e.editor.session.getLine(row);
+  const breakpoints = e.editor.session.getBreakpoints();
+  if (
+    breakpoints[row] === undefined &&
+    content.length !== 0 &&
+    !content.includes('//') &&
+    !content.includes('debugger;')
+  ) {
+    e.editor.session.setBreakpoint(row, undefined!);
+  } else {
+    e.editor.session.clearBreakpoint(row);
   }
+  e.stop();
+  handleEditorUpdateBreakpoints(e.editor.session.getBreakpoints());
+};
 
-  public componentDidUpdate(prevProps: EditorProps) {
-    const newCursorPosition = this.props.newCursorPosition;
-    if (newCursorPosition && newCursorPosition !== prevProps.newCursorPosition) {
-      this.moveCursor(newCursorPosition);
+const makeHandleAnnotationChange = (session: any) => () => {
+  const annotations = session.getAnnotations();
+  let count = 0;
+  for (const anno of annotations) {
+    if (anno.type === 'info') {
+      anno.type = 'error';
+      anno.className = 'ace_error';
+      count++;
     }
   }
+  if (count !== 0) {
+    session.setAnnotations(annotations);
+  }
+};
 
-  public getMarkers = () => {
-    const markerProps: IAceEditorProps['markers'] = [];
-    for (const lineNum of this.props.highlightedLines) {
-      markerProps.push({
-        startRow: lineNum[0],
-        startCol: 0,
-        endRow: lineNum[1],
-        endCol: 1,
-        className: 'myMarker',
-        type: 'fullLine'
-      });
-    }
-    return markerProps;
-  };
-
-  // chapter selector used to choose the correct source mode
-  public chapterNo = () => {
-    let chapter = this.props.sourceChapter;
-    let variant = this.props.sourceVariant;
-    let external = this.props.externalLibraryName;
-    if (chapter === undefined) {
-      chapter = 1;
-    }
-    if (variant === undefined) {
-      variant = 'default';
-    }
-    if (external === undefined) {
-      external = 'NONE';
+const makeCompleter = (handlePromptAutocomplete: DispatchProps['handlePromptAutocomplete']) => ({
+  getCompletions: (editor: any, session: any, pos: any, prefix: any, callback: any) => {
```

Comment:
It would be ideal if we could type all these function arguments here, but it would require digging a little into the Ace types, which are poorly documented. @blackening do you want me to help out with this?

Reply:
Oh nice, the types work out. Addressed in b5a249b.